File: cpp/dnd-character/dnd_character.cpp
#include "dnd_character.h"
#include <algorithm>
#include <random>
#include <cmath>

namespace dnd_character
{
    static std::mt19937 &rng()
    {
        static std::mt19937 gen{std::random_device{}()};
        return gen;
    }

    static int roll_4d6_drop_lowest()
    {
        std::uniform_int_distribution<int> dist(1, 6);

        std::vector<int> dice(4);
        for (int &d : dice)
            d = dist(rng());

        std::sort(dice.begin(), dice.end(), std::greater<int>());

        return dice[0] + dice[1] + dice[2];
    }

    int ability()
    {
        return roll_4d6_drop_lowest();
    }

    int modifier(int score)
    {
        return (int)std::floor((score - 10) / 2.0);
    }

    Character::Character() : strength(ability()),
                             dexterity(ability()),
                             constitution(ability()),
                             intelligence(ability()),
                             wisdom(ability()),
                             charisma(ability()),
                             hitpoints(10 + modifier(constitution))
    {
    }

} // namespace dnd_character

File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.Comparator;
import java.util.List;
import java.util.Random;

class DnDCharacter {

    private final int strength = ability(rollDice());
    private final int dexterity = ability(rollDice());
    private final int constitution = ability(rollDice());
    private final int intelligence = ability(rollDice());
    private final int wisdom = ability(rollDice());
    private final int charisma = ability(rollDice());
    private final int hitpoints = 10 + modifier(constitution);

    int ability(List<Integer> scores) {
        return scores.stream()
                .sorted(Comparator.reverseOrder())
                .limit(3)
                .reduce(0, Integer::sum);
    }

    List<Integer> rollDice() {
        final var rnd = new Random();
        return rnd.ints(4, 1, 7).boxed().toList();
    }

    int modifier(int input) {
        return (int) Math.floor((input - 10) / 2.0);
    }

    int getStrength() {
        return strength;
    }

    int getDexterity() {
        return dexterity;
    }

    int getConstitution() {
        return constitution;
    }

    int getIntelligence() {
        return intelligence;
    }

    int getWisdom() {
        return wisdom;
    }

    int getCharisma() {
        return charisma;
    }

    int getHitpoints() {
        return hitpoints;
    }
}
